torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -452,6 +452,25 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   terminateProcessGroup_.store(true);
   watchdogCV_.notify_one();
   workListCV_.notify_one();
+
+  std::unique_lock<std::mutex> lock(workListMutex_);
+  // TODO: We can potentially merge this functionality into the workCleanup
+  // thread or just allow the destructor to free workList_.
+  // Clean up any remaining items in the workList_ instead of waiting for the
+  // workCleanup Thread to be scheduled again.
+  for (auto it = workList_.begin(); it != workList_.end();
+       /* no increment*/) {
+    auto& work = *it;
+    if (work->isCompleted()) {
+      it = workList_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  // Wait for workList_ to become empty before proceeding with shutdown.
+  workListCV_.wait(lock, [&]() -> bool { return workList_.empty(); });
+  lock.unlock();
```

Comment:
With above workList cleanup, we won't need to wait?

Reply:
We're still blocking to ensure the workList is empty, so the workCleanupThread will continue looping and removing works when they are completed.